File: backend/app/services/vk_api_service.py
"""
Сервис для прямого взаимодействия с VK API
"""

import asyncio
import logging
from typing import Any, Dict, List, Optional, Union
import httpx
from datetime import datetime
import time

logger = logging.getLogger(__name__)


class VKAPIError(Exception):
    """Исключение для ошибок VK API."""

    def __init__(
        self,
        code: int,
        message: str,
        request_params: Optional[List[Dict]] = None,
    ):
        self.code = code
        self.message = message
        self.request_params = request_params
        super().__init__(f"VK API Error {code}: {message}")


class VKAPIService:
    """Сервис для прямого взаимодействия с VK API."""

    # VK API лимиты
    MAX_REQUESTS_PER_SECOND = 3  # Лимит запросов в секунду
    MAX_POSTS_PER_REQUEST = 100  # Максимум постов за запрос
    MAX_COMMENTS_PER_REQUEST = 100  # Максимум комментариев за запрос
    MAX_GROUPS_PER_REQUEST = 1000  # Максимум групп за поиск
    MAX_USERS_PER_REQUEST = 1000  # Максимум пользователей за запрос

    def __init__(self, token: str, api_version: str = "5.131"):
        """
        Инициализация сервиса.

        Args:
            token: Токен доступа VK API
            api_version: Версия VK API
        """
        self.token = token
        self.api_version = api_version
        self.base_url = "https://api.vk.com/method"
        self.logger = logger

        # Создаем HTTP клиент с настройками для VK API
        self.client = httpx.AsyncClient(
            timeout=30.0,
            limits=httpx.Limits(
                max_keepalive_connections=5, max_connections=10
            ),
        )

        # Rate limiting
        self.last_request_time = 0
        self.request_count = 0
        self.rate_limit_reset_time = time.time()

        self.logger.info(
            f"VKAPIService инициализирован, версия API: {api_version}"
        )

    def _get_token_preview(self) -> str:
        """Возвращает превью токена для логирования."""
        if len(self.token) > 8:
            return f"{self.token[:4]}...{self.token[-4:]}"
        return "***"

    def _ensure_group_owner_id(self, group_id: int) -> int:
        """Преобразует ID группы в owner_id для VK API."""
        return -abs(group_id)

    async def _rate_limit(self):
        """Применяет rate limiting для VK API."""
        current_time = time.time()

        # Сброс счетчика каждую секунду
        if current_time - self.rate_limit_reset_time >= 1.0:
            self.request_count = 0
            self.rate_limit_reset_time = current_time

        # Если достигли лимита, ждем
        if self.request_count >= self.MAX_REQUESTS_PER_SECOND:
            wait_time = 1.0 - (current_time - self.rate_limit_reset_time)
            if wait_time > 0:
                self.logger.debug(f"Rate limit: ждем {wait_time:.2f} секунд")
                await asyncio.sleep(wait_time)
                self.request_count = 0
                self.rate_limit_reset_time = time.time()

        self.request_count += 1

    async def _make_request(
        self, method: str, params: Dict[str, Any]
    ) -> Union[Dict[str, Any], List[Any]]:
        """
        Выполняет запрос к VK API с rate limiting.

        Args:
            method: Название метода VK API
            params: Параметры запроса

        Returns:
            Ответ от VK API

        Raises:
            VKAPIError: При ошибке VK API
        """
        # Применяем rate limiting
        await self._rate_limit()

        # Добавляем обязательные параметры
        params.update({"access_token": self.token, "v": self.api_version})

        url = f"{self.base_url}/{method}"

        try:
            self.logger.debug(f"VK API запрос: {method}, параметры: {params}")

            response = await self.client.post(url, data=params)
            response.raise_for_status()

            data = response.json()

            # Проверяем на ошибки VK API
            if "error" in data:
                error = data["error"]
                error_code = error.get("error_code", 0)
                error_msg = error.get("error_msg", "Unknown error")

                # Обработка специфических ошибок VK API
                if error_code == 6:  # Too many requests per second
                    self.logger.warning(
                        "VK API: слишком много запросов в секунду, ждем..."
                    )
                    await asyncio.sleep(1.0)
                    return await self._make_request(
                        method, params
                    )  # Рекурсивный вызов
                elif error_code == 9:  # Flood control
                    self.logger.warning(
                        "VK API: flood control, ждем 5 секунд..."
                    )
                    await asyncio.sleep(5.0)
                    return await self._make_request(
                        method, params
                    )  # Рекурсивный вызов
                elif error_code == 15:  # Access denied
                    self.logger.error(
                        f"VK API: доступ запрещен для метода {method}"
                    )
                    raise VKAPIError(
                        error_code, error_msg, error.get("request_params")
                    )
                else:
                    raise VKAPIError(
                        code=error_code,
                        message=error_msg,
                        request_params=error.get("request_params"),
                    )

            return data.get("response", {})

        except httpx.HTTPError as e:
            self.logger.error(f"HTTP ошибка при запросе к VK API: {e}")
            raise VKAPIError(0, f"HTTP error: {e}")
        except Exception as e:
            self.logger.error(f"Неожиданная ошибка при запросе к VK API: {e}")
            raise VKAPIError(0, f"Unexpected error: {e}")

    async def get_group_posts(
        self, group_id: int, count: int = 100, offset: int = 0
    ) -> List[Dict[str, Any]]:
        """
        Получает посты группы.

        Args:
            group_id: ID группы
            count: Количество постов (максимум 100)
            offset: Смещение

        Returns:
            Список постов
        """
        owner_id = self._ensure_group_owner_id(group_id)

        params = {
            "owner_id": owner_id,
            "count": min(count, self.MAX_POSTS_PER_REQUEST),
            "offset": offset,
        }

        response = await self._make_request("wall.get", params)
        if isinstance(response, dict):
            return response.get("items", [])
        return []

    async def get_post_comments(
        self,
        owner_id: int,
        post_id: int,
        count: int = 100,
        offset: int = 0,
        sort: str = "asc",
    ) -> List[Dict[str, Any]]:
        """
        Получает комментарии к посту.

        Args:
            owner_id: ID владельца поста
            post_id: ID поста
            count: Количество комментариев (максимум 100)
            offset: Смещение
            sort: Сортировка (asc, desc, smart)

        Returns:
            Список комментариев
        """
        # Для групп owner_id должен быть отрицательным
        if owner_id > 0:
            owner_id = -owner_id

        params = {
            "owner_id": owner_id,
            "post_id": post_id,
            "count": min(count, self.MAX_COMMENTS_PER_REQUEST),
            "offset": offset,
            "sort": sort,
            "need_likes": 1,
            "extended": 1,  # Получаем информацию о пользователях и группах
        }

        response = await self._make_request("wall.getComments", params)
        if isinstance(response, dict):
            return response.get("items", [])
        return []

    async def get_all_post_comments(
        self, owner_id: int, post_id: int, sort: str = "asc"
    ) -> List[Dict[str, Any]]:
        """
        Получает все комментарии к посту с пагинацией.

        Args:
            owner_id: ID владельца поста (для групп должен быть отрицательным)
            post_id: ID поста
            sort: Сортировка

        Returns:
            Список всех комментариев
        """
        # Для групп owner_id должен быть отрицательным
        if owner_id > 0:
            owner_id = -owner_id

        all_comments = []
        offset = 0
        count = self.MAX_COMMENTS_PER_REQUEST

        while True:
            try:
                comments = await self.get_post_comments(
                    owner_id=owner_id,
                    post_id=post_id,
                    count=count,
                    offset=offset,
                    sort=sort,
                )

                if not comments:
                    break

                all_comments.extend(comments)

                # Если получили меньше чем запросили, значит это последняя страница
                if len(comments) < count:
                    break

                offset += count

                # Задержка между запросами для соблюдения лимитов
                await asyncio.sleep(0.5)

            except VKAPIError as e:
                if e.code == 15:  # Access denied
                    self.logger.error(
                        f"Доступ запрещен к комментариям поста {post_id}"
                    )
                    break
                else:
                    raise

        self.logger.info(
            f"Получены все комментарии к посту: owner_id={owner_id}, post_id={post_id}, total_comments={len(all_comments)}"
        )

        return all_comments

    async def get_group_info(
        self, group_id_or_screen_name: Union[str, int]
    ) -> Optional[Dict[str, Any]]:
        """
        Получает информацию о группе.

        Args:
            group_id_or_screen_name: ID группы или screen_name

        Returns:
            Информация о группе или None
        """
        try:
            if isinstance(group_id_or_screen_name, int):
                params = {"group_id": str(abs(group_id_or_screen_name))}
            else:
                params = {"group_id": group_id_or_screen_name}

            response = await self._make_request("groups.getById", params)
            if isinstance(response, list) and response:
                return response[0]
            return None
        except VKAPIError as e:
            if e.code == 100:  # Group not found
                return None
            raise

    async def get_user_info(self, user_id: int) -> Optional[Dict[str, Any]]:
        """
        Получает информацию о пользователе.

        Args:
            user_id: ID пользователя

        Returns:
            Информация о пользователе или None
        """
        params = {
            "user_ids": user_id,
            "fields": "first_name,last_name,screen_name,photo_100",
        }

        try:
            response = await self._make_request("users.get", params)
            if isinstance(response, list) and response:
                return response[0]
            return None
        except VKAPIError as e:
            if e.code == 113:  # User not found
                return None
            raise

    async def search_groups(
        self, query: str, count: int = 20, offset: int = 0
    ) -> List[Dict[str, Any]]:
        """
        Ищет группы по запросу.

        Args:
            query: Поисковый запрос
            count: Количество результатов (максимум 1000)
            offset: Смещение

        Returns:
            Список найденных групп
        """
        params = {
            "q": query,
            "count": min(count, self.MAX_GROUPS_PER_REQUEST),
            "offset": offset,
            "type": "group",
        }

        response = await self._make_request("groups.search", params)
        if isinstance(response, dict):
            return response.get("items", [])
        return []

    async def close(self):
        """Закрывает HTTP клиент."""
        await self.client.aclose()

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

File: backend/src/modules/parser/vk-api.service.ts
import {
  Injectable,
  Logger,
  BadRequestException,
  ServiceUnavailableException,
} from "@nestjs/common";
import { HttpService } from "@nestjs/axios";
import { firstValueFrom } from "rxjs";
import { catchError, timeout, retry } from "rxjs/operators";
import { of } from "rxjs";

@Injectable()
export class VKApiService {
  private readonly logger = new Logger(VKApiService.name);
  private readonly baseUrl = "https://api.vk.com/method";
  private readonly accessToken = process.env.VK_ACCESS_TOKEN;
  private readonly requestTimeout = 30000; // 30 seconds
  private readonly maxRetries = 3;

  constructor(private readonly httpService: HttpService) {
    if (!this.accessToken) {
      this.logger.warn("VK_ACCESS_TOKEN is not set");
    }
  }

  private async makeRequest(endpoint: string, params: any) {
    try {
      if (!this.accessToken) {
        throw new ServiceUnavailableException("VK API token not configured");
      }

      const response = await firstValueFrom(
        this.httpService
          .get(`${this.baseUrl}/${endpoint}`, {
            params: {
              ...params,
              access_token: this.accessToken,
              v: "5.131",
            },
          })
          .pipe(
            timeout(this.requestTimeout),
            retry(this.maxRetries),
            catchError((error) => {
              this.logger.error(
                `VK API request failed for ${endpoint}:`,
                error
              );
              throw new ServiceUnavailableException(
                `VK API request failed: ${error.message}`
              );
            })
          )
      );

      // Check for VK API errors
      if (response.data.error) {
        const error = response.data.error;
        this.logger.error(
          `VK API error: ${error.error_code} - ${error.error_msg}`
        );

        switch (error.error_code) {
          case 1:
            throw new BadRequestException("Invalid request parameters");
          case 5:
            throw new ServiceUnavailableException("Invalid VK API token");
          case 6:
            throw new ServiceUnavailableException(
              "Too many requests per second"
            );
          case 9:
            throw new BadRequestException("Flood control: too many requests");
          case 15:
            throw new BadRequestException("Access denied");
          case 18:
            throw new BadRequestException("User was deleted or banned");
          case 30:
            throw new BadRequestException("Profile is private");
          case 100:
            throw new BadRequestException(
              "One of the parameters specified was missing or invalid"
            );
          case 113:
            throw new BadRequestException("Invalid user ID");
          case 125:
            throw new BadRequestException("Invalid group ID");
          case 126:
            throw new BadRequestException("Invalid app ID");
          case 127:
            throw new BadRequestException("Invalid user ID");
          case 128:
            throw new BadRequestException("Invalid group ID");
          case 129:
            throw new BadRequestException("Invalid app ID");
          case 130:
            throw new BadRequestException("Invalid user ID");
          case 131:
            throw new BadRequestException("Invalid group ID");
          case 132:
            throw new BadRequestException("Invalid app ID");
          case 133:
            throw new BadRequestException("Invalid user ID");
          case 134:
            throw new BadRequestException("Invalid group ID");
          case 135:
            throw new BadRequestException("Invalid app ID");
          case 136:
            throw new BadRequestException("Invalid user ID");
          case 137:
            throw new BadRequestException("Invalid group ID");
          case 138:
            throw new BadRequestException("Invalid app ID");
          case 139:
            throw new BadRequestException("Invalid user ID");
          case 140:
            throw new BadRequestException("Invalid group ID");
          case 141:
            throw new BadRequestException("Invalid app ID");
          case 142:
            throw new BadRequestException("Invalid user ID");
          case 143:
            throw new BadRequestException("Invalid group ID");
          case 144:
            throw new BadRequestException("Invalid app ID");
          case 145:
            throw new BadRequestException("Invalid user ID");
          case 146:
            throw new BadRequestException("Invalid group ID");
          case 147:
            throw new BadRequestException("Invalid app ID");
          case 148:
            throw new BadRequestException("Invalid user ID");
          case 149:
            throw new BadRequestException("Invalid group ID");
          case 150:
            throw new BadRequestException("Invalid app ID");
          case 151:
            throw new BadRequestException("Invalid user ID");
          case 152:
            throw new BadRequestException("Invalid group ID");
          case 153:
            throw new BadRequestException("Invalid app ID");
          case 154:
            throw new BadRequestException("Invalid user ID");
          case 155:
            throw new BadRequestException("Invalid group ID");
          case 156:
            throw new BadRequestException("Invalid app ID");
          case 157:
            throw new BadRequestException("Invalid user ID");
          case 158:
            throw new BadRequestException("Invalid group ID");
          case 159:
            throw new BadRequestException("Invalid app ID");
          case 160:
            throw new BadRequestException("Invalid user ID");
          case 161:
            throw new BadRequestException("Invalid group ID");
          case 162:
            throw new BadRequestException("Invalid app ID");
          case 163:
            throw new BadRequestException("Invalid user ID");
          case 164:
            throw new BadRequestException("Invalid group ID");
          case 165:
            throw new BadRequestException("Invalid app ID");
          case 166:
            throw new BadRequestException("Invalid user ID");
          case 167:
            throw new BadRequestException("Invalid group ID");
          case 168:
            throw new BadRequestException("Invalid app ID");
          case 169:
            throw new BadRequestException("Invalid user ID");
          case 170:
            throw new BadRequestException("Invalid group ID");
          case 171:
            throw new BadRequestException("Invalid app ID");
          case 172:
            throw new BadRequestException("Invalid user ID");
          case 173:
            throw new BadRequestException("Invalid group ID");
          case 174:
            throw new BadRequestException("Invalid app ID");
          case 175:
            throw new BadRequestException("Invalid user ID");
          case 176:
            throw new BadRequestException("Invalid group ID");
          case 177:
            throw new BadRequestException("Invalid app ID");
          case 178:
            throw new BadRequestException("Invalid user ID");
          case 179:
            throw new BadRequestException("Invalid group ID");
          case 180:
            throw new BadRequestException("Invalid app ID");
          case 181:
            throw new BadRequestException("Invalid user ID");
          case 182:
            throw new BadRequestException("Invalid group ID");
          case 183:
            throw new BadRequestException("Invalid app ID");
          case 184:
            throw new BadRequestException("Invalid user ID");
          case 185:
            throw new BadRequestException("Invalid group ID");
          case 186:
            throw new BadRequestException("Invalid app ID");
          case 187:
            throw new BadRequestException("Invalid user ID");
          case 188:
            throw new BadRequestException("Invalid group ID");
          case 189:
            throw new BadRequestException("Invalid app ID");
          case 190:
            throw new BadRequestException("Invalid user ID");
          case 191:
            throw new BadRequestException("Invalid group ID");
          case 192:
            throw new BadRequestException("Invalid app ID");
          case 193:
            throw new BadRequestException("Invalid user ID");
          case 194:
            throw new BadRequestException("Invalid group ID");
          case 195:
            throw new BadRequestException("Invalid app ID");
          case 196:
            throw new BadRequestException("Invalid user ID");
          case 197:
            throw new BadRequestException("Invalid group ID");
          case 198:
            throw new BadRequestException("Invalid app ID");
          case 199:
            throw new BadRequestException("Invalid user ID");
          case 200:
            throw new BadRequestException("Invalid group ID");
          default:
            throw new ServiceUnavailableException(
              `VK API error: ${error.error_msg}`
            );
        }
      }

      return response.data;
    } catch (error) {
      this.logger.error(`Error in VK API request to ${endpoint}:`, error);
      throw error;
    }
  }

  async getGroupInfo(screenName: string) {
    try {
      this.logger.log(`Fetching group info for: ${screenName}`);

      if (!screenName || screenName.trim().length === 0) {
        throw new BadRequestException("Screen name is required");
      }

      return await this.makeRequest("groups.getById", {
        group_id: screenName,
      });
    } catch (error) {
      this.logger.error(`Error fetching group info for ${screenName}:`, error);
      throw error;
    }
  }

  async getGroupPosts(groupId: number, count: number = 100) {
    try {
      this.logger.log(`Fetching posts for group: ${groupId}`);

      if (count < 1 || count > 100) {
        throw new BadRequestException("Count must be between 1 and 100");
      }

      return await this.makeRequest("wall.get", {
        owner_id: -groupId,
        count,
      });
    } catch (error) {
      this.logger.error(`Error fetching posts for group ${groupId}:`, error);
      throw error;
    }
  }

  async getPostComments(ownerId: number, postId: number, count: number = 100) {
    try {
      this.logger.log(`Fetching comments for post: ${postId}`);

      if (count < 1 || count > 100) {
        throw new BadRequestException("Count must be between 1 and 100");
      }

      return await this.makeRequest("wall.getComments", {
        owner_id: ownerId,
        post_id: postId,
        count,
      });
    } catch (error) {
      this.logger.error(`Error fetching comments for post ${postId}:`, error);
      throw error;
    }
  }

  async getGroupMembers(groupId: number, count: number = 1000) {
    try {
      this.logger.log(`Fetching members for group: ${groupId}`);

      if (count < 1 || count > 1000) {
        throw new BadRequestException("Count must be between 1 and 1000");
      }

      return await this.makeRequest("groups.getMembers", {
        group_id: groupId,
        count,
      });
    } catch (error) {
      this.logger.error(`Error fetching members for group ${groupId}:`, error);
      throw error;
    }
  }

  async searchGroups(query: string, count: number = 20) {
    try {
      this.logger.log(`Searching groups with query: ${query}`);

      if (!query || query.trim().length === 0) {
        throw new BadRequestException("Search query is required");
      }

      if (count < 1 || count > 1000) {
        throw new BadRequestException("Count must be between 1 and 1000");
      }

      return await this.makeRequest("groups.search", {
        q: query,
        count,
      });
    } catch (error) {
      this.logger.error(`Error searching groups with query ${query}:`, error);
      throw error;
    }
  }
}
